File: .test/ziwei-test/lib/ziwei/models/gan_zhi.rb
module Ziwei
  module Models
    class GanZhi
      attr_reader :stem, :branch

      def initialize(args = {})
        @stem = args[:stem] || args["stem"] || :giap
        @stem = :giap if @stem == :_

        @branch = args[:branch] || args["branch"] || :ty
        @branch = :ty if @branch == :_

        stem_index = Configs::Stems::Orders.index(@stem)
        branch_index = Configs::Branches::Orders.index(@branch)

        raise "Invalid stem-branch pair" if (stem_index.odd? ^ branch_index.odd?)
      end

      def inspect
        [
          Configs::Stems::Names[@stem],
          Configs::Branches::Names[@branch]
        ].join(" ")
      end
    end
  end
end

File: ziwei-js/src/models/gan_zhi.js
_Ziwei_Models_GanZhi = class {
  constructor(args = {}) {
    var _stem = args['stem'] || 'giap';
    var _branch = args['branch'] || 'ty';

    if (_stem === undefined || _branch === undefined) {
      throw "Undefined stem or branch";
      return;
    }

    this.stem = _stem;
    this.branch = _branch;

    var stemIndex = Ziwei.Configs.Stems.Orders.indexOf(this.stem);
    var branchIndex = Ziwei.Configs.Branches.Orders.indexOf(this.branch);

    if (stemIndex < 0 || branchIndex < 0)
      throw "Invalid stem or branch";

    if (stemIndex.isEven() ^ branchIndex.isEven() === true)
      throw "Invalid stem-branch pair";
  };

  toString() {
    return [
      Ziwei.Configs.Stems.Names[this.stem],
      Ziwei.Configs.Branches.Names[this.branch]
    ].join(" ");
  };
};